src/frontend/Errors.ml

```diff
@@ -22,21 +22,25 @@ type t =
   | Semantic_error of Semantic_error.t
   | DebugDataError of (Middle.Location_span.t * string)
 
-let get_context_callback ?code Middle.Location.{filename; included_from; _} () =
-  (* By the time we are printing an error, all these files are already resolved. *)
+let get_context ?code Middle.Location.{filename; included_from; _} () =
+  (* If the location is not included from anywhere, it's the orignal code *)
+  let code = if Option.is_none included_from then code else None in
+  (* Otherwise, by the time we are printing an error,
+     all these files are already resolved. *)
   match !Include_files.include_provider with
```

Comment:
Both branches of the `match` map the option, let's just do an early return.
```ocaml
if Option.is_none included_from
  then String.split_lines code
  else match ...
```

Reply:
`code` is a `string option`, not just a `string`

Reply:
And if `code` is None this function fails to produce context for stdin source.

Looks like that can only happen if the pretty printer produces incorrect code:
https://github.com/stan-dev/stanc3/blob/528c72c26e3fe466545ba0bf29eac262d7fb69b2/src/frontend/Pretty_print_prog.ml#L56-L58
I don't think this has ever quite worked. (It always prints empty context; `In_channel.read_lines` probably hides a file-not-found error.) That function should use `Errors.pp` instead of `Errors.to_string`.

The only other place with `Errors.pp ~code:None` is
https://github.com/stan-dev/stanc3/blob/528c72c26e3fe466545ba0bf29eac262d7fb69b2/src/stanc/stanc.ml#L78-L80
and that might be better served by exposing a `Errors.pp_debug_data_error`.

Reply:
We also don't set it in the call for file inputs, a few lines down from there:
https://github.com/stan-dev/stanc3/blob/528c72c26e3fe466545ba0bf29eac262d7fb69b2/src/stanc/stanc.ml#L86

Reply:
Right, don't know how I missed that.
I don't think I would want to pass the source as `'File|'Code` (I'm not a fan of it in `Parse.parse` either) so this code should stay as is.

I think changing the pretty printer to use `Errors.pp ~code:pretty` is still worth it, even though it is an obscure ICE.

Reply:
> 'm not a fan of it in Parse.parse either

Do you have something you prefer? (I suppose it would be a minor improvement to make it an actual type rather than a polymorphic variant?)

Reply:
I do not. If I had, I would have mentioned it when you added it (I reviewed that pull request too and it's only been a month since)